<a href="https://colab.research.google.com/github/PeggyK1/DS-Unit-1-Sprint-2-Statistics/blob/master/Peggy_Krom__LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [0]:
# YOUR WORK HERE

In [1]:
#Loading Data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-06-09 19:31:22--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.06s   

2020-06-09 19:31:22 (321 KB/s) - ‘student.zip’ saved [20478/20478]



In [2]:
#Unzipping
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [0]:
import pandas as pd
from scipy import stats

In [5]:
df = pd.read_csv('student-mat.csv', sep=';')
print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


**Null Hypothesis:** Data is due to chance and there is no relationship between the two categories

**Alternative Hypothesis:** Data is *not* due to chance and theri *is* a relationship between the two categories

**Confidence Level:** 95%

In [0]:
# YOUR WORK HERE

In [0]:
#Testing Health and Absences Relationship

In [6]:
df['health'].value_counts().sort_index()

1     47
2     45
3     91
4     66
5    146
Name: health, dtype: int64

In [7]:
df['absences'].value_counts().sort_index()

0     115
1       3
2      65
3       8
4      53
5       5
6      31
7       7
8      22
9       3
10     17
11      3
12     12
13      3
14     12
15      3
16      7
17      1
18      5
19      1
20      4
21      1
22      3
23      1
24      1
25      1
26      1
28      1
30      1
38      1
40      1
54      1
56      1
75      1
Name: absences, dtype: int64

In [8]:
contingency = pd.crosstab(df['health'], df['absences'] )
contingency

absences  0   1   2   3   4   5   6   7   ...  26  28  30  38  40  54  56  75
health                                    ...                                
1         13   0   9   0   6   1   4   2  ...   0   0   0   0   0   0   1   0
2         12   0   3   1  11   1   3   0  ...   1   0   0   0   0   0   0   0
3         24   2  15   5   9   0   5   1  ...   0   0   0   0   1   0   0   0
4         16   0  10   1  13   1   9   1  ...   0   0   0   1   0   0   0   0
5         50   1  28   1  14   2  10   3  ...   0   1   1   0   0   1   0   1

[5 rows x 34 columns]

In [9]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("Chi2 Statistic:", chi2)
print("P Value:", p_value)
print("Degrees of Freedom:", dof)
print("Expected Frequencyies Table", expected)

Chi2 Statistic: 138.6370339526242
P Value: 0.3289894307147289
Degrees of Freedom: 132
Expected Frequencyies Table [[13.6835443   0.35696203  7.73417722  0.95189873  6.30632911  0.59493671
   3.68860759  0.83291139  2.61772152  0.35696203  2.02278481  0.35696203
   1.4278481   0.35696203  1.4278481   0.35696203  0.83291139  0.11898734
   0.59493671  0.11898734  0.47594937  0.11898734  0.35696203  0.11898734
   0.11898734  0.11898734  0.11898734  0.11898734  0.11898734  0.11898734
   0.11898734  0.11898734  0.11898734  0.11898734]
 [13.10126582  0.34177215  7.40506329  0.91139241  6.03797468  0.56962025
   3.53164557  0.79746835  2.50632911  0.34177215  1.93670886  0.34177215
   1.36708861  0.34177215  1.36708861  0.34177215  0.79746835  0.11392405
   0.56962025  0.11392405  0.4556962   0.11392405  0.34177215  0.11392405
   0.11392405  0.11392405  0.11392405  0.11392405  0.11392405  0.11392405
   0.11392405  0.11392405  0.11392405  0.11392405]
 [26.49367089  0.69113924 14.97468354  1.843

**Conclusion:** Based on the conventional confidence level of 95%,

With a Chi2 Stat of 138 and a p-value of .32

I **fail to reject** the Null Hypothesis, that the categories are not related. I do not have enough information to say that the data is not due to chance.

In [0]:
#Testing Going Out and Workday Alcohol Cunsumption Relationship

In [10]:
df['goout'].value_counts().sort_index()

1     23
2    103
3    130
4     86
5     53
Name: goout, dtype: int64

In [11]:
df['Dalc'].value_counts().sort_index()

1    276
2     75
3     26
4      9
5      9
Name: Dalc, dtype: int64

In [12]:
contingency2 = pd.crosstab(df['goout'], df['Dalc'])
contingency2

Dalc    1   2  3  4  5
goout                 
1      20   2  1  0  0
2      81  16  3  2  1
3      99  22  6  1  2
4      50  23  9  4  0
5      26  12  7  2  6

In [13]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency2)

print("Chi2 Statistic:", chi2)
print("P Value:", p_value)
print("Degrees of Freedom:", dof)
print("Expected Frequencyies Table", expected)

Chi2 Statistic: 48.78628219209503
P Value: 3.5714751508089954e-05
Degrees of Freedom: 16
Expected Frequencyies Table [[16.07088608  4.36708861  1.51392405  0.52405063  0.52405063]
 [71.96962025 19.55696203  6.77974684  2.34683544  2.34683544]
 [90.83544304 24.6835443   8.55696203  2.96202532  2.96202532]
 [60.09113924 16.32911392  5.66075949  1.95949367  1.95949367]
 [37.03291139 10.06329114  3.48860759  1.20759494  1.20759494]]


**Conclusion:** Based on the conventional confidence level of 95%,

With a Chi2 Stat of 48.7 and a p-value of 3.57e-05

I **reject** the Null Hypothesis and suggest the Alternative, that the two categories are related

In [0]:
#Testing Health and Workday Alcohol Consumption (Dalc)

In [15]:
contingency3 = pd.crosstab(df['health'], df['Dalc'])
contingency3

Dalc     1   2   3  4  5
health                  
1       34   8   3  1  1
2       31  10   4  0  0
3       72  11   4  3  1
4       47  13   0  2  4
5       92  33  15  3  3

In [16]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency3)

print("Chi2 Statistic:", chi2)
print("P Value:", p_value)
print("Degrees of Freedom:", dof)
print("Expected Frequencyies Table", expected)

Chi2 Statistic: 21.60477822244893
P Value: 0.1564172504736364
Degrees of Freedom: 16
Expected Frequencyies Table [[ 32.84050633   8.92405063   3.09367089   1.07088608   1.07088608]
 [ 31.44303797   8.5443038    2.96202532   1.02531646   1.02531646]
 [ 63.58481013  17.27848101   5.98987342   2.07341772   2.07341772]
 [ 46.1164557   12.53164557   4.3443038    1.50379747   1.50379747]
 [102.01518987  27.72151899   9.61012658   3.32658228   3.32658228]]


**Conclusion:** Based on the conventional confidence level of 95%,

With a Chi2 Stat of 21.6 and a p-value of .156

I **fail to reject** the Null Hypothesis, that the categories are not related. I do not have enough information to say that the data is not due to chance.

In [0]:
#Testing Travel Time and Study Time Relationship

In [17]:
contingency4 = pd.crosstab(df['studytime'], df['traveltime'])
contingency4

traveltime    1   2   3  4
studytime                 
1            60  31  11  3
2           133  51  11  3
3            47  16   1  1
4            17   9   0  1

In [18]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency4)

print("Chi2 Statistic:", chi2)
print("P Value:", p_value)
print("Degrees of Freedom:", dof)
print("Expected Frequencyies Table", expected)

Chi2 Statistic: 11.200170929144969
P Value: 0.26223746329595793
Degrees of Freedom: 9
Expected Frequencyies Table [[ 68.3164557   28.44303797   6.11392405   2.12658228]
 [128.82531646  53.63544304  11.52911392   4.01012658]
 [ 42.29113924  17.60759494   3.78481013   1.3164557 ]
 [ 17.56708861   7.31392405   1.5721519    0.54683544]]


**Conclusion:** Based on the conventional confidence level of 95%,

With a Chi2 Stat of 11.2 and a p-value of .26

I **fail to reject** the Null Hypothesis, that the categories are not related. I do not have enough information to say that the data is not due to chance.

In [0]:
#Testing Workday Alcohol Consumption and Weekend Alcohol Consumption Relationship

In [19]:
contingency5 = pd.crosstab(df['Dalc'], df['Walc'])
contingency5

Walc    1   2   3   4  5
Dalc                    
1     150  65  42  15  4
2       1  18  29  22  5
3       0   1   8  11  6
4       0   1   1   3  4
5       0   0   0   0  9

In [20]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency5)

print("Chi2 Statistic:", chi2)
print("P Value:", p_value)
print("Degrees of Freedom:", dof)
print("Expected Frequencyies Table", expected)

Chi2 Statistic: 287.0019095823893
P Value: 1.2457721336028707e-51
Degrees of Freedom: 16
Expected Frequencyies Table [[105.50886076  59.39240506  55.89873418  35.63544304  19.56455696]
 [ 28.67088608  16.13924051  15.18987342   9.6835443    5.3164557 ]
 [  9.93924051   5.59493671   5.26582278   3.35696203   1.84303797]
 [  3.44050633   1.93670886   1.82278481   1.16202532   0.63797468]
 [  3.44050633   1.93670886   1.82278481   1.16202532   0.63797468]]


**Conclusion:** Based on the conventional confidence level of 95%,

With a Chi2 Stat of 287 and a p-value of 1.2e-51

I **reject** the Null Hypothesis and suggest the Alternative, that the two categories are related

**Dependent:** 
1. Going Out and Workday Alcohol Consumption 
2. Workday Alcohol Consumption and Weekend Alcohol Consumption

**Independent:**
1. Health and Absences
2. Health and Workday Alcohol Consumption
3. Travel Time and Study Time

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [0]:
# YOUR WORK HERE

In [0]:
import numpy as np

In [0]:
#Manual Test of Dalc and Walc Relationship

In [38]:
Observed = pd.crosstab(df['Dalc'], df['Walc'], margins=True)
Observed

Walc    1   2   3   4   5  All
Dalc                          
1     150  65  42  15   4  276
2       1  18  29  22   5   75
3       0   1   8  11   6   26
4       0   1   1   3   4    9
5       0   0   0   0   9    9
All   151  85  80  51  28  395

In [50]:
(276*151)/395

105.50886075949367

In [39]:
Observed.shape

(6, 6)

In [42]:
#Observed
observed=contingency5.values
observed

array([[150,  65,  42,  15,   4],
       [  1,  18,  29,  22,   5],
       [  0,   1,   8,  11,   6],
       [  0,   1,   1,   3,   4],
       [  0,   0,   0,   0,   9]])

In [43]:
row_totals = df['Walc'].value_counts().sort_index().values 
row_totals

array([151,  85,  80,  51,  28])

In [44]:
col_totals = df['Dalc'].value_counts().sort_index().values
col_totals

array([276,  75,  26,   9,   9])

In [45]:
sample_size = df.shape[0]
sample_size

395

In [51]:
my_list = []

for i, row_total in enumerate(row_totals):
  row = []
  for j, column_total in enumerate(col_totals):
    expected_val  = ((column_total*row_total) / sample_size)
    row.append(expected_val)
  my_list.append(row)

my_list

[[105.50886075949367,
  28.670886075949365,
  9.939240506329114,
  3.440506329113924,
  3.440506329113924],
 [59.392405063291136,
  16.139240506329113,
  5.594936708860759,
  1.9367088607594938,
  1.9367088607594938],
 [55.89873417721519,
  15.189873417721518,
  5.265822784810126,
  1.8227848101265822,
  1.8227848101265822],
 [35.63544303797468,
  9.683544303797468,
  3.356962025316456,
  1.1620253164556962,
  1.1620253164556962],
 [19.564556962025318,
  5.3164556962025316,
  1.8430379746835444,
  0.6379746835443038,
  0.6379746835443038]]

In [53]:
expected = np.array(my_list).T
expected

array([[105.50886076,  59.39240506,  55.89873418,  35.63544304,
         19.56455696],
       [ 28.67088608,  16.13924051,  15.18987342,   9.6835443 ,
          5.3164557 ],
       [  9.93924051,   5.59493671,   5.26582278,   3.35696203,
          1.84303797],
       [  3.44050633,   1.93670886,   1.82278481,   1.16202532,
          0.63797468],
       [  3.44050633,   1.93670886,   1.82278481,   1.16202532,
          0.63797468]])

In [55]:
chi2 = (((observed-expected)**2) / expected).sum()
chi2

287.00190958238926

In [0]:
#Double Checking with Scipy

In [56]:
chi2, p_value, dof, expected = stats.chi2_contingency(observed)

print("Chi2 Statistic:", chi2)
print("P Value:", p_value)
print("Degrees of Freedom:", dof)
print("Expected Frequencyies Table", expected)

Chi2 Statistic: 287.0019095823893
P Value: 1.2457721336028707e-51
Degrees of Freedom: 16
Expected Frequencyies Table [[105.50886076  59.39240506  55.89873418  35.63544304  19.56455696]
 [ 28.67088608  16.13924051  15.18987342   9.6835443    5.3164557 ]
 [  9.93924051   5.59493671   5.26582278   3.35696203   1.84303797]
 [  3.44050633   1.93670886   1.82278481   1.16202532   0.63797468]
 [  3.44050633   1.93670886   1.82278481   1.16202532   0.63797468]]


## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [0]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

1. **Sample Estimate:** Chosen number of observations to make inferences about a population. In the real world, sample size is usually determined by cost, time, or convenience
2. **Hypothesis Test:** An observation that is testable by comparing two datasets or comparing one dataset to a set data example. 
They are useful in determining how two categories or variables are related to each other.
3. **Null Hypothesis:** A type of hypothesis in statistics that assumes there is no difference between two variables
4. **P-Value = Probability Value:** It represents the probability of calculating test results as extreme as the observed results. Calculating p-value is subtracting the confidence level % from 100%
5. **Statistical Significance:** When the null hypothesis suggests that a result is rare
6. **Test Statistic:** Used to test statistical hypothesis. The test statistic is the difference between the sample mean and the population mean divided by the standard deviation/sample size. The larger the ttest the smaller the p-value
7. **Degrees of Freedom:** Number of values in final calculation that are free to very.
**1 Sample** Number of categories - 1
**2 Sample** (Number of Crosstab Rows - 1)*(Number of Crosstab Columns-1)


## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)